### PR TITLE
feat: BREAKING request/response; ILQP packets

### PR DESCRIPTION
**Pull request by sentientwaffle**
* Depends on: {five-bells-shared,ilp{,-plugin-bells,-routing}}#dj-ilqp
* Add `connector.registerRequestHandler()` (used by ilp-kit to add some request handling).
* Use plugin request/response semantics.
* Use ILQP packets (supports quotes by source/destination amount, but not liquidity curve).

### CODE REVIEW

**Comment by michielbdejong** [comment]
also related to https://github.com/interledger/rfcs/pull/198 and https://github.com/interledgerjs/ilp-plugin-virtual/pull/62 for request/response, and https://github.com/interledger/rfcs/blob/master/0008-interledger-quoting-protocol/0008-interledger-quoting-protocol.md#example-request should be updated as well as part of this refactor.

Also, depending on what we decide to do with https://github.com/interledgerjs/ilp-connector/issues/344#issuecomment-300619614, this refactor could require a "version 3" of ilp-kit, and then all nodes in the open Interledger network would need to update at the same time, coordinated with switching over connector.land to the new format. However, if we decide that remote quoting is only currently used for default routes, then technically the change is still breaking for ilp-kit, but the impact would be restricted to compatibility between two ilp-kits where one acts as the default route of the other, so that would not require a coordinated network upgrade.

File: src/lib/route-broadcaster.js [hunk 1]
```diff
@@ -243,9 +242,10 @@ class RouteBroadcaster {
       const connector = configRoute.connectorAccount
       const targetPrefix = configRoute.targetPrefix
 
+      const maxUint = new BigNumber(2).pow(64).minus(1).toString()
```
Comment (emschwartz) [inline src/lib/route-broadcaster.js]: This should just be a constant in here

File: src/lib/route-builder.js [hunk 2]
```diff
@@ -144,32 +147,25 @@ class RouteBuilder {
       ilpPacket.account, ilpPacket.amount)
 
     const sourceLedger = sourceTransfer.ledger
-    // Use `findBestHopForSourceAmount` since the source amount includes the slippage.
     const nextHop = this.routingTables.findBestHopForSourceAmount(
       sourceLedger, ilpPacket.account, sourceTransfer.amount)
     if (!nextHop) {
       log.info('could not find route for source transfer: ' + JSON.stringify(sourceTransfer))
-      log.debug('current routing tables (simplified to 10 points): ' + JSON.stringify(this.routingTables.toJSON(10)))
```
Comment (emschwartz) [inline src/lib/route-builder.js]: Why are you taking this out?

**Comment by michielbdejong** [comment]
I created mj-test-dj-ilqp branches on the following repos
* five-bells-shared
* five-bells-ledger <- no changes here, but it should use the changes in five-bells-shared
* five-bells-integration-test
* ilp-connector
* ilp-routing
* ilp-plugin-bells
* ilp-plugin-virtual
* ilp-kit <- no changes here, but it should use the changes in all components
* ilp

File: src/lib/route-builder.js [hunk 3]
```diff
@@ -2,97 +2,100 @@
...
-const IlpError = require('../errors/ilp-error')
+const IncomingTransferError = require('../errors/incoming-transfer-error')
 const getDeterministicUuid = require('../lib/utils').getDeterministicUuid
 const log = require('../common/log').create('route-builder')
 const startsWith = require('lodash/startsWith')
 
 class RouteBuilder {
   /**
-   * @param {RoutingTables} routingTables
    * @param {Ledgers} ledgers
```
Comment (michielbdejong) [inline src/lib/route-builder.js]: add `@param quoter` here

File: src/lib/quoter.js [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+   * @param {Integer} request.destinationHoldDuration
+   * @returns {Object}
+   */
+  * quoteByDestinationAmount (request) {
+    if (request.destinationAmount === '0') {
+      throw new InvalidAmountSpecifiedError('destinationAmount must be positive')
+    }
+    return yield this._quoteByAmount(request,
+      IlpPacket.serializeIlqpByDestinationRequest,
+      IlpPacket.deserializeIlqpByDestinationResponse)
```
Comment (michielbdejong) [inline src/lib/quoter.js]: `* _quoteByAmount (request)` only takes one argument, not three.

File: src/lib/quoter.js [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+   * @param {Integer} request.destinationHoldDuration
+   * @returns {Object}
+   */
+  * quoteBySourceAmount (request) {
+    if (request.sourceAmount === '0') {
+      throw new InvalidAmountSpecifiedError('sourceAmount must be positive')
+    }
+    return yield this._quoteByAmount(request,
+      IlpPacket.serializeIlqpBySourceRequest,
+      IlpPacket.deserializeIlqpBySourceResponse)
```
Comment (michielbdejong) [inline src/lib/quoter.js]: `* _quoteByAmount (request)` only takes one argument, not three.

**Comment by michielbdejong** [comment]
Do the unit tests test remote quoting? I'm running into ['MockPlugin.sendRequest is not implemented'](https://github.com/interledgerjs/ilp-connector/blob/1eab9ec5cc8e554eb167f2931080985f99df7237/test/mocks/mockPlugin.js#L32)

**Comment by sentientwaffle** [comment]
@michielbdejong the unit tests pass for me. Can you send me the stack, and/or which test in particular is failing?

**Comment by emschwartz** [comment]
@sentientwaffle @michielbdejong Should this be merged now? I think integration tests are failing because some of these changes have been merged but not all

**Comment by michielbdejong** [comment]
Yes, I'll have a look.

**Pull request merged**